File: small-problems/medium-2/triangle_sides.rb
# Triangle Sides
#
# Problem     |---------------------------------------------------------------|
#             |Given three numbers, return the type of triangle formed with the
#             |length of the sides equal to the numbers given, or invalid if a
#             |triangle may not be formed with those lengths.
#       Input |Three numbers.
#       Output|A symbol, :equilateral , :isosceles , :scalene , or :invalid
#       Edges |Zero sides
#       Rules |Equilateral triangles have sides with equal lengths.
#             |Isosceles have two sides with equal length and one different.
#             |Scalene triangles have all different lengths.
#             |The sum of the two shorter sides must be greater than the longest
#             |  side in order for the triangle to be valid.
#             |All lengths must be greater than 0.
#   Questions |
# Example     |---------------------------------------------------------------|
# triangle(3, 3, 3) == :equilateral
# triangle(3, 3, 1.5) == :isosceles
# triangle(3, 4, 5) == :scalene
# triangle(0, 3, 3) == :invalid
# triangle(3, 1, 1) == :invalid
# Data        |---------------------------------------------------------------|
#             |N/A
# Algorithm   |---------------------------------------------------------------|
#             |Return :invalid if any side is 0 or less.
#             |Create an array containing the side lengths and sort it.
#             |Return :equilateral if calling uniq on the array equals one.
#             |Return :isosceles if calling uniq on the array equals two and 
#             |  the sum of the first two elements in the array is greater than
#             |  the third element.
#             |Return :scalene if the sum of the first two elements exceeds the
#             |  third
#             |Return :invalid
# Code________|_______________________________________________________________|
#
def triangle(side1, side2, side3)
  return :invalid if side1 <= 0 || side2 <= 0 || side3 <= 0
  sides = [side1, side2, side3].sort
  same = sides.uniq.size
  return :equilateral if same == 1
  valid = sides[0] + sides[1] > sides[2]
  return :isosceles if same == 2 && valid
  return :scalene if valid
  :invalid
end

puts triangle(3, 3, 3) == :equilateral
puts triangle(3, 3, 1.5) == :isosceles
puts triangle(3, 4, 5) == :scalene
puts triangle(0, 3, 3) == :invalid
puts triangle(3, 1, 1) == :invalid

File: LS215/small_problems/medium_2/triangle_sides.js
'use strict';

// the location may have to be altered within each distinct folder
const TESTER = require('../../tester');
const t = TESTER.makeTestSuite();

/*
Problem-->

-- IO
---- Input  -> Three numbers representing three sides of a theoretical triangle.
---- Output -> A string denoting the type of triangle formed by the numbers given, or 'invalid' if the triangle is not valid.
---- Edges  -> Not enough sides.

-- Rules
Valid triangles have:
- Sides greater than 0.
- The sum of the two shorter sides must exceed the longest side in length.

Triangle Types:
Equilateral - All sides are equal in length.
Isosceles   - Two sides are equal in length.
Scalene     - All sides differ in length.

-- Notes

Examples-->
---- Tests for Invalid Input
t.addTest(triangle, 'invalid', '', '', '');
t.addTest(triangle, 'invalid', 0, 4, 7);
t.addTest(triangle, 'invalid', 0, 0, 0);
t.addTest(triangle, 'invalid', 0);
---- Tests for Normal Operation
t.addTest(triangle, expected, args);
t.addTest(triangle, 'equilateral', 3, 3, 3);
t.addTest(triangle, 'isosceles', 3, 3, 1.5);
t.addTest(triangle, 'scalene', 3, 4, 5);
t.addTest(triangle, 'invalid', 0, 3, 3);
t.addTest(triangle, 'invalid', 3, 1, 1);
t.addTest(triangle, 'equilateral', 1, 1, 1);
t.addTest(triangle, 'scalene', 4, 7, 9);
t.addTest(triangle, 'isosceles', 9, 4, 9);

Data Structure-->
We will be using arrays at one point to calculate the number of unique values.

Algorithm-->
Create an object which holds the three return values mapped to the keys of the number of unique sides.
Filter out falsy values as edges.
If the sides do not combine in terms of length, return invalid.
Return the value at the String key value of the number of unique sides to the triangle.

✓Functions-->
[✓] unique(arr)
----- Returns a new array containing all of the unique values from the original array.
[✓] validLengthCombination(sides)
----- Returns true if the two smallest values in the sides array are greater than the largest value when summed.
*/

const DEBUG = false;

// Helpers
const trianglesByUniqueSideCount = {
  '1': 'equilateral',
  '2': 'isosceles',
  '3': 'scalene'
};

const sum = (a, b) => a + b;
const validLengthCombination = (sides) => {
  sides = sides.slice().sort();
  const shortestSum = sides.slice(0, 2).reduce(sum);
  return shortestSum > Math.max(...sides);
}

const unique = (arr) => {
  const result = [];
  arr.forEach((em) => {
    if (!result.includes(em)) result.push(em);
  });
  return result;
}

// Edges
const edgeCheck = (s1, s2, s3) => {
  if (!s1 || !s2 || !s3) return 'invalid';
  return null;
}

// Primary
function triangle(s1, s2, s3) {
  const isEdge = edgeCheck(s1, s2, s3);
  if (isEdge !== null) return isEdge;
  
  const sides = [s1, s2, s3];
  if (!validLengthCombination(sides)) return 'invalid';
  return trianglesByUniqueSideCount[unique(sides).length];
}

if (DEBUG) {
  t.addTest(unique, [], []);                     // empty array
  t.addTest(unique, [1], [1]);                   // non-empty array
  t.addTest(unique, [1], [1, 1]);                // duplicates removed
  t.addTest(unique, [1, 2, 3, 4, 5], [1, 2, 3, 4, 5, 4, 2]); // only one of each
  t.addTest(validLengthCombination, true,  [3, 3, 3]);
  t.addTest(validLengthCombination, false, [3, 2, 1]);
  t.addTest(validLengthCombination, true,  [3, 4, 5]);
  t.addTest(validLengthCombination, true,  [3, 3, 1.78]);
} else {
  t.addTest(triangle, 'invalid', '', '', '');
  t.addTest(triangle, 'invalid', 0, 4, 7);
  t.addTest(triangle, 'invalid', 0, 0, 0);
  t.addTest(triangle, 'invalid', 0);
  t.addTest(triangle, 'equilateral', 3, 3, 3);
  t.addTest(triangle, 'isosceles', 3, 3, 1.5);
  t.addTest(triangle, 'scalene', 3, 4, 5);
  t.addTest(triangle, 'invalid', 0, 3, 3);
  t.addTest(triangle, 'invalid', 3, 1, 1);
  t.addTest(triangle, 'equilateral', 1, 1, 1);
  t.addTest(triangle, 'scalene', 4, 7, 9);
  t.addTest(triangle, 'isosceles', 9, 4, 9);
}

// Tests
t.runSuite();
